Sources/AsyncHTTPClient/HTTPHandler.swift

```diff
@@ -481,18 +481,65 @@ extension HTTPClient {
             let channel: Channel? = self.lock.withLock {
                 if !cancelled {
                     cancelled = true
-                    return self.channel
+                    return self.connection?.channel
                 }
                 return nil
             }
             channel?.triggerUserOutboundEvent(TaskCancelEvent(), promise: nil)
```

Comment:
Looks like there is a risk that the `TaskCancelEvent()` isn't going to be passed to the pipeline if a connection wasn't available yet. It also seems like accesses to `self.connection` should have proper synchronisation.

Reply:
Fixed